File: scala/2015/day03b/Program.scala
import scala.collection.immutable.HashSet
import scala.io.Source
import scala.util.Using
import scala.sys.exit

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def process(filename: String): Int = {
    val content = Using.resource(Source.fromFile(filename))(_.mkString)
    content.toList.foldLeft(((0, 0), (0, 0), HashSet[(Int, Int)](), false)) {
      (acc : ((Int, Int), (Int, Int), HashSet[(Int, Int)], Boolean), ch : Char) =>
        val (santa, roboSanta, positions, santaMove) = acc
        val nextPosition = santaMove match {
          case true => ch match {
            case '^' => (santa._1, santa._2 + 1)
            case 'v' => (santa._1, santa._2 - 1)
            case '<' => (santa._1 - 1, santa._2)
            case '>' => (santa._1 + 1, santa._2)
            case _ => santa
          }
          case false => ch match {
            case '^' => (roboSanta._1, roboSanta._2 + 1)
            case 'v' => (roboSanta._1, roboSanta._2 - 1)
            case '<' => (roboSanta._1 - 1, roboSanta._2)
            case '>' => (roboSanta._1 + 1, roboSanta._2)
            case _ => roboSanta
          }
        }
        val nextSanta = if ( santaMove ) nextPosition else santa
        val nextRoboSanta = if ( santaMove ) roboSanta else nextPosition
        val nextPositions = positions + nextPosition
        val nextMove = !santaMove
        (nextSanta, nextRoboSanta, nextPositions, nextMove)
    }._3.size
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case filename :: Nil =>
        val result = process(filename)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}

File: fsharp/2015/day03b/Program.fs
open System
open System.Collections.Generic
open System.IO

open AOC_Utils.Geometry

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processFile(filename: string) =
    let mutable positions = HashSet<Position2D<int>>()
    let f (santa: Position2D<int>, roboSanta: Position2D<int>, santaMove: bool) ch =
        if santaMove then
            let santa' = match ch with
                         | '^' -> new Position2D<int>(santa.X, santa.Y + 1)
                         | 'v' -> new Position2D<int>(santa.X, santa.Y - 1)
                         | '<' -> new Position2D<int>(santa.X - 1, santa.Y)
                         | '>' -> new Position2D<int>(santa.X + 1, santa.Y)
                         | _ -> santa
            positions.Add(santa') |> ignore
            (santa', roboSanta, false)
        else
            let roboSanta' = match ch with
                             | '^' -> new Position2D<int>(roboSanta.X, roboSanta.Y + 1)
                             | 'v' -> new Position2D<int>(roboSanta.X, roboSanta.Y - 1)
                             | '<' -> new Position2D<int>(roboSanta.X - 1, roboSanta.Y)
                             | '>' -> new Position2D<int>(roboSanta.X + 1, roboSanta.Y)
                             | _ -> roboSanta
            positions.Add(roboSanta') |> ignore
            (santa, roboSanta', true)

    try
        File.ReadAllText(filename)
        |> Seq.fold f (Position2D<int>(0, 0), Position2D<int>(0, 0), false)
        |> ignore
        positions.Count
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
